core/src/main/scala/org/apache/spark/scheduler/Task.scala

```diff
@@ -155,7 +155,13 @@ private[spark] abstract class Task[T](
    */
   def collectAccumulatorUpdates(taskFailed: Boolean = false): Seq[AccumulatorV2[_, _]] = {
     if (context != null) {
-      context.taskMetrics.accumulators().filter { a => !taskFailed || a.countFailedValues }
+      context.taskMetrics.internalAccums.filter { a =>
+        // RESULT_SIZE accumulator is always zero at executor, we need to send it back as its
+        // value will be updated at driver side.
+        !a.isZero || a.name == Some(InternalAccumulator.RESULT_SIZE)
+      // zero value external accumulators may still be useful, e.g. SQLMetrics, we should not filter
```

Comment:
It's task level statistics. Let's say an operator launched 100 tasks to execute it, and 99 tasks don't spill, only one task spills 100 mb, then the avg of `spilling size` will be 100 mb, if we don't include zero values. This is obviously wrong.